### PR TITLE
KAFKA-8215: Upgrade Rocks to v6.0.1

### PR DESCRIPTION
**Pull request by ableegoldman**
Users have indicated that the memory usage of RocksDB can be excessive, in particular as it scales with the number of rocksdb instances meaning it could cause OOM if a thread ends up with a large number of stores following a rebalance. 

Upgrading Rocks exposes a number of new configs, including the WriteBufferManager which --along with the existing TableFormatConfigs -- allows users to limit the **total** memory allocated to Rocks across all instances in a single process

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ableegoldman** [comment]
Looks like facebook finally got sick of me bugging them to publish the latest RocksJava ;)

@cadonna @mjsax @vvcephei @abbccdda @guozhangwang @bbejeck

**Comment by mjsax** [comment]
Can we run systems test for this update?

**Comment by ableegoldman** [comment]
> Can we run systems test for this update?

Ack

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:1412]: Do we need to expose this setting?  As of right now, we don't expose RocksDB ttl functionality.

**Review by bbejeck — APPROVED** [review]
Thanks for the PR @ableegoldman, just one minor comment otherwist LGTM pending system test/build test results.

**Comment by bbejeck** [comment]
Java 8 passed, Java 11 failed with ` /home/jenkins/tools/gradle/4.10.2/bin/gradle: No such file or directory`

retest this please

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:1412]: RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest requires us to wrap every Options method -- but I suppose we could log a warning that this will be ignored?

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:1412]: SGTM

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:380]: It is orthogonal to this PR, but I've some personal opinions on whether Streams eat this complexity of maintaining the wrapper (and hence each time its upgraded we need to make changes like this , or we should just push this burden to users to distinguish DBOptions and CFOptions): https://issues.apache.org/jira/browse/KAFKA-7869

@mjsax has expressed his preference clearly :) and I'd like to hear other's thoughts: if people favor the first option, then I can go ahead and close that ticket.

@cadonna @ableegoldman @bbejeck @vvcephei @abbccdda

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: `CompactRangeOptions extends RocksObject`, we need to explicitly `close()` it otherwise it's memory leak.

Since all such objects are created the same, how about just keeping a single copy of it and close it during `store.close()` similar to other xxOptions?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Ditto for LRUCache: `Cache extends RocksObject`. We need to close it explicitly.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:380]: Hm. On the one hand this did feel like rather a lot of trivial work (although if we try to keep up with Rocks releases as we go in the future it wouldn't be so bad :) ) and it's not that hard to figure out, especially using an IDE with a good autocomplete.
I'm actually not too concerned about putting the burden on the users, but on the other hand imagine someone has an existing app where they set a large number of options using the ConfigSetter. They'd have to go back and change all of them by hand...personally I'd be a bit annoyed.

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:380]: Actually I just thought of something: if the user constructs a new RocksObject and passes it to options we will never close it and thus leak memory :(

I see two options for us:
1)) use the wrapper to capture all objects created in ConfigSetter and close them ourselves RocksDBStore#close
2) add a #close() method to the ConfigSetter interface and tell users to close any new objects themselves

I think 2) sounds far more reasonable for us to manage but I am worried users won't see this and continue to leak. We'd probably want to cherry-pick this fix back to older branches..?
WDYT @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Good catch. I found that the BloomFilter also needs to be closed (and isn't). Should this be fixed in older versions?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Ack

**Comment by apurvam** [comment]
What testing have we done to ensure that the memory is actually bounded the way we expect? Also it would be good to update the docs with a simple example of how to set this config and bound total memory used by rocksDB in a single process.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Let's wait until the KIP is merged before this one.

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Actually we got the BloomFilter fix into the 2.2.1 release :)

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Ah right, I think then it is not blocked on anything (the KIP was semi-orthogonal to this).

**Comment by guozhangwang** [comment]
I think it is a bit hard to write a unit test to make sure the memory for block cache is bounded. What we can do though, is to have an one-off benchmark that gradually add more state stores into a dummy topology, and measure the JVM memory with v.s. w/o this PR, and see if that increased differently maybe? @ableegoldman

**Comment by ableegoldman** [comment]
I agree I'm not sure how one would get this into a unit test, but I'm working on a simple streams app that should demonstrate whether or not it's working. Unfortunately it's off-heap memory so we can't just look at the JVM usage

**Comment by apurvam** [comment]
Sorry, I didn't mean a unit test. I think a manual test which clearly shows the memory bounds with and without this change in rebalance scenario is enough. Documenting the test and the observed results is desirable.

**Comment by ableegoldman** [comment]
Closing this in favor of [minor version bump](https://github.com/apache/kafka/pull/6743)

**Pull request closed without merging**
